wenet/bin/export_onnx.py

```diff
@@ -0,0 +1,263 @@
+        super().__init__()
+        self.decoder = decoder
+        self.reverse_weight = reverse_weight
+        self.beam_size = beam_size
+
+    def forward(self,
+                encoder_out: torch.Tensor,
+                encoder_lens: torch.Tensor,
+                hyps_pad_sos: torch.Tensor,
+                hyps_lens: torch.Tensor,
+                r_hyps_pad_sos: torch.Tensor):
+        """Encoder
+        Args:
+            encoder_out: B x T x F
+            encoder_lens: B
+            hyps_pad_sos: B x beam x T2,
+                        hyps with sos and padded by ignore id
+            hyps_lens: B x beam, length for each hyp with sos
+            r_hyps_pad_sos: B x beam x T2,
+                    reversed hyps with sos and padded by ignore id
+        Returns:
+            decoder_out: B x beam x T2 x V
+            r_decoder_out: B x beam x T2 x V
+        """
+        B, T, F = encoder_out.shape
+        encoder_out = encoder_out.repeat(1, self.beam_size, 1).view(-1, T, F)
+        encoder_mask = ~make_pad_mask(encoder_lens, T).unsqueeze(1)
+        encoder_mask = encoder_mask.repeat(1, self.beam_size, 1).view(-1, 1, T)
+        T2 = hyps_pad_sos.shape[2]
+        B2 = B * self.beam_size
+        hyps_pad = hyps_pad_sos.view(B2, T2)
+        hyps_lens = hyps_lens.view(B2,)
+        r_hyps_pad = r_hyps_pad_sos.view(B2, T2)
+        decoder_out, r_decoder_out, _ = self.decoder(
+            encoder_out, encoder_mask, hyps_pad, hyps_lens, r_hyps_pad,
+            self.reverse_weight)
+        decoder_out = torch.nn.functional.log_softmax(decoder_out, dim=-1)
+        V = decoder_out.shape[-1]
+        decoder_out = decoder_out.view(B, self.beam_size, T2, V)
+        r_decoder_out = torch.nn.functional.log_softmax(r_decoder_out, dim=-1)
+        r_decoder_out = r_decoder_out.view(B, self.beam_size, T2, V)
+        return decoder_out, r_decoder_out
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='export your script model')
+    parser.add_argument('--config', required=True, help='config file')
+    parser.add_argument('--checkpoint', required=True, help='checkpoint model')
+    parser.add_argument('--beam_size', default=10, type=int, required=False,
+                        help="beam size would be ctc output size")
+    parser.add_argument('--output_onnx_dir',
+                        default="onnx_model",
+                        help='output onnx encoder and decoder directory')
+    args = parser.parse_args()
+
+    torch.manual_seed(0)
+    torch.set_printoptions(precision=10)
+
+    with open(args.config, 'r') as fin:
+        configs = yaml.load(fin, Loader=yaml.FullLoader)
+
+    configs["encoder_conf"]["use_dynamic_chunk"] = False
+    model = init_asr_model(configs)
+    load_checkpoint(model, args.checkpoint)
+    model.eval()
+
+    bz = 32
+    seq_len = 100
+    beam_size = args.beam_size
+    feature_size = configs["collate_conf"]["feature_extraction_conf"]["mel_bins"]
```

Comment:
better to use configs['input_dim'] instead of this.

Reply:
Got it.